src/app/clusters/thread-border-router-management-server/thread-br-mgmt-server.cpp

```diff
@@ -0,0 +1,358 @@
+ *    limitations under the License.
+ */
+
+#include "thread-br-mgmt-server.h"
+
+#include "app-common/zap-generated/cluster-objects.h"
+#include "app-common/zap-generated/ids/Attributes.h"
+#include "app-common/zap-generated/ids/Clusters.h"
+#include "app-common/zap-generated/ids/Commands.h"
+#include "app/AttributeAccessInterfaceRegistry.h"
+#include "app/AttributeValueEncoder.h"
+#include "app/CommandHandler.h"
+#include "app/CommandHandlerInterface.h"
+#include "app/InteractionModelEngine.h"
+#include "app/MessageDef/StatusIB.h"
+#include "app/clusters/general-commissioning-server/general-commissioning-server.h"
+#include "app/data-model/Nullable.h"
+#include "app/server/Server.h"
+#include "lib/core/CHIPError.h"
+#include "lib/support/CodeUtils.h"
+#include "lib/support/Span.h"
+#include "lib/support/ThreadOperationalDataset.h"
+#include "platform/CHIPDeviceEvent.h"
+#include "platform/PlatformManager.h"
+#include "protocols/interaction_model/StatusCode.h"
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ThreadBorderRouterManagement {
+
+using Protocols::InteractionModel::Status;
+
+static bool CheckOverCASESession(CommandHandlerInterface::HandlerContext & ctx)
+{
+    Messaging::ExchangeContext * exchangeCtx = ctx.mCommandHandler.GetExchangeContext();
+    if (!exchangeCtx || !exchangeCtx->HasSessionHandle() || !exchangeCtx->GetSessionHandle()->IsSecureSession() ||
+        exchangeCtx->GetSessionHandle()->AsSecureSession()->GetSecureSessionType() != Transport::SecureSession::Type::kCASE)
+    {
+        ChipLogError(Zcl, "This command MUST be over a valid CASE session");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::UnsupportedAccess);
+        return false;
+    }
+    return true;
+}
+
+static bool CheckFailSafeArmed(CommandHandlerInterface::HandlerContext & ctx)
+{
+    auto & failSafeContext = Server::GetInstance().GetFailSafeContext();
+    if (failSafeContext.IsFailSafeArmed(ctx.mCommandHandler.GetAccessingFabricIndex()))
+    {
+        return true;
+    }
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::FailsafeRequired);
+    return false;
+}
+
+static bool CheckDelegate(CommandHandlerInterface::HandlerContext & ctx, Delegate * delegate)
+{
+    if (!delegate)
+    {
+        ChipLogError(Zcl, "Thread Border Router Management server not initialized");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidInState);
+    }
+    return delegate;
+}
+
+void ServerInstance::HandleGetDatasetRequest(HandlerContext & ctx, Delegate::DatasetType type)
+{
+    VerifyOrReturn(CheckOverCASESession(ctx));
+    VerifyOrReturn(CheckDelegate(ctx, mDelegate));
+
+    Commands::DatasetResponse::Type response;
+    Thread::OperationalDataset dataset;
+    CHIP_ERROR err = mDelegate->GetDataset(dataset, type);
+    if (err != CHIP_NO_ERROR)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, StatusIB(err).mStatus);
```

Comment:
This uses "automatic" CHIP_ERROR conversion which is not clean or well specified today. This likely will cause the implementation to return confusing status codes that don't meet spec in some conditions.

Reply:
Maked this matches the SPEC